packages/capacitor-plugin/android/src/main/kotlin/com/capacitorjs/plugins/filesystem/FilesystemPlugin.kt

```diff
@@ -0,0 +1,356 @@
+    private val controller: IONFILEController by lazy { IONFILEController(context.applicationContext) }
+
+    override fun handleOnDestroy() {
+        super.handleOnDestroy()
+        coroutineScope.cancel()
+    }
+
+    @PluginMethod
+    fun readFile(call: PluginCall) {
+        val input: ReadFileOptions = call.getReadFileOptions() ?: run {
+            call.sendError(FilesystemErrors.invalidInputMethod(call.methodName))
+            return
+        }
+        runWithPermission(input.uri, call) { uri ->
+            controller.readFile(uri, input.options)
+                .onSuccess { call.sendSuccess(result = createReadResultObject(it)) }
+                .onFailure { call.sendError(it.toFilesystemError(call.methodName)) }
+        }
+    }
+
+    @PluginMethod(returnType = PluginMethod.RETURN_CALLBACK)
+    fun readFileInChunks(call: PluginCall) {
+        val input: ReadFileInChunksOptions = call.getReadFileInChunksOptions() ?: run {
+            call.sendError(FilesystemErrors.invalidInputMethod(call.methodName))
+            return
+        }
+        runWithPermission(input.uri, call) { uri ->
+            controller.readFileInChunks(uri, input.options)
+                .onEach { chunk ->
+                    call.sendSuccess(result = createReadResultObject(chunk), keepCallback = true)
+                }
+                .onCompletion { error ->
+                    if (error == null) {
+                        call.sendSuccess(result = createReadResultObject(""))
+                    }
+                }
+                .catch {
+                    call.sendError(it.toFilesystemError(call.methodName))
+                }
+                .launchIn(coroutineScope)
+        }
+    }
+
+    @PluginMethod
+    fun writeFile(call: PluginCall) {
+        val input: WriteFileOptions = call.getWriteFileOptions() ?: run {
+            call.sendError(FilesystemErrors.invalidInputMethod(call.methodName))
+            return
+        }
+        runWithPermission(input.uri, call) { uri ->
+            controller.saveFile(uri, input.options)
+                .onSuccess { uriSaved ->
+                    // update mediaStore index only if file was written to external storage
+                    if (uri.inExternalStorage) {
+                        uriSaved.path?.let {
+                            MediaScannerConnection.scanFile(context, arrayOf(it), null, null)
+                        }
+                    }
+                    call.sendSuccess(result = createWriteResultObject(uriSaved, input.options.mode))
+                }
+                .onFailure { call.sendError(it.toFilesystemError(call.methodName)) }
+        }
+    }
+
+    @PluginMethod
+    fun appendFile(call: PluginCall) {
+        try {
+            call.data.putOpt(INPUT_APPEND, true)
+        } catch (ex: JSONException) {
+            Log.w(logTag, "Tried to set `append` in `PluginCall`, but got exception", ex)
```

Comment:
Should this be `Log.w` or `Log.e`? Just asking because if we're returning an error, maybe it would make sense to log it to the error channel.